File: src/cleansia_android/core/src/main/java/cz/cleansia/core/consent/SignupConsentClient.kt
package cz.cleansia.core.consent

import cz.cleansia.core.network.ApiError
import cz.cleansia.core.network.ApiResult

/**
 * Per-app binding seam for the GDPR consent endpoints, mirroring
 * [cz.cleansia.core.notifications.DeviceRegistrationClient]: each app implements it
 * over its own OpenAPI-generated `GdprApi` so the parking and delivery rules live
 * once, in [SignupConsentRepository].
 */
interface SignupConsentClient {

    /**
     * Every consent type the account has already answered — granted **or withdrawn**.
     * Null when the read itself failed, which is not the same as "answered nothing".
     */
    suspend fun answeredTypes(): Set<SignupConsentType>?

    suspend fun grant(type: SignupConsentType): ConsentGrantOutcome
}

enum class ConsentGrantOutcome {
    Recorded,

    /** The backend refused a duplicate. The record it refuses to duplicate is the one we came to write. */
    AlreadyOnFile,

    Failed,
}

fun ApiResult<*>.toConsentGrantOutcome(): ConsentGrantOutcome = when (this) {
    is ApiResult.Success -> ConsentGrantOutcome.Recorded
    is ApiResult.Error ->
        if (error.namesConsentAlreadyGranted()) ConsentGrantOutcome.AlreadyOnFile else ConsentGrantOutcome.Failed
}

/**
 * Read off [ApiError.BadRequest.errorKey] — the first VALUE in the ProblemDetails `errors` bag.
 * `CleansiaApiController.CreateProblemDetails` keys that bag by `Error.Code` (the offending field,
 * `ConsentType`) and values it with `Error.Message` (the business key), which is the same slot the
 * web and iOS clients resolve their translations from. Matching the bag KEY or the `type` instead
 * would be matching the field name.
 */
private fun ApiError.namesConsentAlreadyGranted(): Boolean {
    val badRequest = this as? ApiError.BadRequest ?: return false
    return badRequest.errorKey == CONSENT_ALREADY_GRANTED
}

private const val CONSENT_ALREADY_GRANTED = "gdpr.consent_already_granted"

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/Consent/SignupConsentClient.swift
import Foundation

/// Per-app binding seam for the GDPR consent endpoints, mirroring `DeviceRegistrationClient`:
/// each app implements it over its own generated Gdpr API so the parking and delivery rules
/// live once, in `SignupConsentRepository`.
public protocol SignupConsentClient: Sendable {
    /// Every consent type the account has already answered — granted **or withdrawn**.
    /// `nil` when the read itself failed, which is not the same as "answered nothing".
    func answeredTypes() async -> Set<SignupConsentType>?

    func grant(_ type: SignupConsentType) async -> ConsentGrantOutcome
}

public enum ConsentGrantOutcome: Equatable, Sendable {
    case recorded

    /// The backend refused a duplicate. The record it refuses to duplicate is the one we came to write.
    case alreadyOnFile

    case failed
}

public extension Result where Failure == ApiError {
    var consentGrantOutcome: ConsentGrantOutcome {
        switch self {
        case .success:
            .recorded
        case let .failure(error):
            error.namesConsentAlreadyGranted ? .alreadyOnFile : .failed
        }
    }
}

private extension ApiError {
    /// Read off the first VALUE in the ProblemDetails `errors` bag, which is what
    /// `ApiError.fromProblemDetails` prefers. `CleansiaApiController.CreateProblemDetails`
    /// keys that bag by `Error.Code` (the offending field, `ConsentType`) and values it with
    /// `Error.Message` (the business key). Matching the bag key or `type` instead would be
    /// matching the field name.
    var namesConsentAlreadyGranted: Bool {
        httpStatus == 400 && code == "gdpr.consent_already_granted"
    }
}
